File: app/src/main/java/com/theradcolor/kernel/activities/gpuActivity.java
package com.theradcolor.kernel.activities;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ContextThemeWrapper;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.grarak.kerneladiutor.utils.Prefs;
import com.theradcolor.kernel.R;
import com.theradcolor.kernel.utils.kernel.GPU;

public class gpuActivity extends AppCompatActivity implements View.OnClickListener{

    LinearLayout governor, gpu_min, gpu_max, gpu_pwr;
    TextView gov, min_freq, max_freq, pwr_lvl;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_gpu);
        gov = findViewById(R.id.tv_gpu_gov);
        min_freq = findViewById(R.id.tv_gpu_min_freq);
        max_freq = findViewById(R.id.tv_gpu_max_freq);
        pwr_lvl = findViewById(R.id.tv_gpu_pwr_lvl);

        governor = findViewById(R.id.ll_gpu_gov);
        governor.setOnClickListener(this);
        gpu_min = findViewById(R.id.ll_gpu_min);
        gpu_min.setOnClickListener(this);
        gpu_max = findViewById(R.id.ll_gpu_max);
        gpu_max.setOnClickListener(this);
        gpu_pwr = findViewById(R.id.ll_gpu_pwr_lvl);
        gpu_pwr.setOnClickListener(this);

        refreshUI();
    }

    /*private void drawGraph() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                for (int i = 0; i < 999; ++i) {
                    String gpuUsage = GPU.getGpuBusy().replaceAll("[-+.^:,%]","");
                    usage.setText(""+GPU.getGpuBusy());
                    curr_freq.setText(GPU.getCurFreq()/1000000+getResources().getString(R.string.mhz));
                    LineChartData data = mChart.getLineChartData();
                    pointValues.add(new PointValue(i, Utils.strToFloat(gpuUsage)));
                    data.getLines().get(0).setValues(new ArrayList<>(pointValues));
                    mChart.setLineChartData(data);
                    setViewport();
                    try {
                        Thread.sleep(500);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        }).start();
    }

    private void setViewport() {
        int size = pointValues.size();
        if (size > maxNumberOfPoints) {
            final Viewport viewport = new Viewport(mChart.getMaximumViewport());
            viewport.left = size - 16;
            mChart.setCurrentViewport(viewport);
        }
    }*/

    public void refreshUI() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                gov.setText(GPU.getGovernor());
                min_freq.setText(GPU.getMinFreq() / GPU.getMinFreqOffset()+getResources().getString(R.string.mhz));
                max_freq.setText(GPU.getMaxFreq() / GPU.getMaxFreqOffset()+getResources().getString(R.string.mhz));
                pwr_lvl.setText(GPU.getPwrLevel());
            }
        }).start();
    }

    @Override
    protected void onResume() {
        super.onResume();
        refreshUI();
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        switch (id){
            case R.id.ll_gpu_gov:
                govDialog();
                break;
            case R.id.ll_gpu_min:
                minDialog();
                break;
            case R.id.ll_gpu_max:
                maxDialog();
                break;
            case R.id.ll_gpu_pwr_lvl:
                pwrDialog();
                break;
        }
    }

    private void govDialog(){
        AlertDialog.Builder builder =  new AlertDialog.Builder(new ContextThemeWrapper(gpuActivity.this, R.style.CustomDialogTheme));
        builder.setTitle("GPU governor");
        String[] GOVs = GPU.getAvailableGovernors().toArray(new String[0]);
        builder.setItems(GOVs, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                GPU.setGovernor(GOVs[which], gpuActivity.this);
                Prefs.saveString("gpu_governor", GOVs[which], gpuActivity.this);
                refreshUI();
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void minDialog(){
        AlertDialog.Builder builder =  new AlertDialog.Builder(new ContextThemeWrapper(gpuActivity.this, R.style.CustomDialogTheme));
        builder.setTitle("GPU minimum freq");
        String[] gpu_min = GPU.getAdjustedFreqs(this).toArray(new String[0]);
        builder.setItems(gpu_min, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                GPU.setMinFreq(GPU.getAvailableFreqs().get(which)/1000000, gpuActivity.this);
                Prefs.saveInt("gpu_min_freq", GPU.getAvailableFreqs().get(which)/1000000, gpuActivity.this);
                refreshUI();
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void maxDialog(){
        AlertDialog.Builder builder =  new AlertDialog.Builder(new ContextThemeWrapper(gpuActivity.this, R.style.CustomDialogTheme));
        builder.setTitle("GPU maximum freq");
        String[] gpu_max = GPU.getAdjustedFreqs(this).toArray(new String[0]);
        builder.setItems(gpu_max, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                GPU.setMaxFreq(GPU.getAvailableFreqs().get(which), gpuActivity.this);
                Prefs.saveInt("gpu_max_freq", GPU.getAvailableFreqs().get(which), gpuActivity.this);
                refreshUI();
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    public void pwrDialog() {
        AlertDialog.Builder builder =  new AlertDialog.Builder(new ContextThemeWrapper(gpuActivity.this, R.style.CustomDialogTheme));
        builder.setTitle("GPU power level");
        final View mpLayout = getLayoutInflater().inflate(R.layout.gpu_pwr_dialog, null);
        EditText editText = mpLayout.findViewById(R.id.pwr_lvl_et);
        editText.setText(GPU.getPwrLevel());
        builder.setPositiveButton("Apply", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                GPU.setPwrLevel(editText.getText().toString(), gpuActivity.this);
                Prefs.saveString("gpu_pwr_lvl", editText.getText().toString(), gpuActivity.this);
                refreshUI();
            }
        });

        builder.setView(mpLayout);
        AlertDialog dialog = builder.create();
        dialog.show();
    }

}

File: app/src/main/java/com/theradcolor/kernel/activities/gpuActivity.kt
package com.theradcolor.kernel.activities

import android.os.Bundle
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ContextThemeWrapper
import com.grarak.kerneladiutor.utils.Prefs
import com.theradcolor.kernel.R
import com.theradcolor.kernel.utils.kernel.GPU
import kotlinx.android.synthetic.main.activity_gpu.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import lecho.lib.hellocharts.model.Line
import lecho.lib.hellocharts.model.LineChartData
import lecho.lib.hellocharts.model.PointValue
import lecho.lib.hellocharts.model.Viewport

class gpuActivity : AppCompatActivity(){

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_gpu)
        val lines:MutableList<Line> = ArrayList()
        val line = Line()
        line.setHasLines(true)
        line.setHasPoints(false)
        line.color = resources.getColor(R.color.colorAccent)
        line.isFilled = true
        lines.add(line)
        val data = LineChartData(lines)
        data.axisXBottom = null
        data.axisYLeft = null
        data.baseValue = java.lang.Float.NEGATIVE_INFINITY
        gpu_chart.lineChartData = data
        refreshUI()
        ll_gpu_gov.setOnClickListener { govDialog() }
        ll_gpu_min.setOnClickListener { minDialog() }
        ll_gpu_max.setOnClickListener { maxDialog() }
        ll_gpu_pwr_lvl.setOnClickListener { pwrDialog() }
        CoroutineScope(Dispatchers.Default).launch {
            readData()
        }
    }

    private var pointValues: MutableList<PointValue> = ArrayList()
    private var maxNumberOfPoints = 16

    private suspend fun readData(){
        for (i in 0..998)
        {
            delay(500)
            gpu_usage.text = GPU.getBusyPer()
            curr_freq.text = GPU.getCurFreq().div(1000000).toString().plus(resources.getString(R.string.mhz))
            val data = gpu_chart.lineChartData
            pointValues.add(PointValue(i.toFloat(), GPU.getBusy().toFloat()))
            data.lines[0].values = ArrayList(pointValues)
            gpu_chart.lineChartData = data
            setViewport()
        }
    }

    private suspend fun setViewport() {
        val size = pointValues.size
        if (size > maxNumberOfPoints) {
            val viewport = Viewport(gpu_chart.maximumViewport)
            viewport.left = (size - 16).toFloat()
            gpu_chart.currentViewport = viewport
        }
    }

    private fun refreshUI() {
        Thread(Runnable {
            tv_gpu_gov!!.text = GPU.getGovernor()
            val minFreq = GPU.getMinFreq().div(GPU.getMinFreqOffset())
            val maxFreq  = GPU.getMaxFreq().div(GPU.getMaxFreqOffset())
            tv_gpu_min_freq.text = (minFreq.toString().plus(resources.getString(R.string.mhz)))
            tv_gpu_max_freq.text = (maxFreq.toString().plus(resources.getString(R.string.mhz)))
            tv_gpu_pwr_lvl!!.text = GPU.getPwrLevel()
        }).start()
    }

    override fun onResume() {
        super.onResume()
        refreshUI()
    }

    private fun govDialog() {
        val builder = AlertDialog.Builder(ContextThemeWrapper(this@gpuActivity, R.style.CustomDialogTheme))
        builder.setTitle("GPU governor")
        val GOVs = GPU.getAvailableGovernors().toTypedArray()
        builder.setItems(GOVs) { dialog, which ->
            GPU.setGovernor(GOVs[which], this@gpuActivity)
            Prefs.saveString("gpu_governor", GOVs[which], this@gpuActivity)
            refreshUI()
        }
        val dialog = builder.create()
        dialog.show()
    }

    private fun minDialog() {
        val builder = AlertDialog.Builder(ContextThemeWrapper(this@gpuActivity, R.style.CustomDialogTheme))
        builder.setTitle("GPU minimum freq")
        val gpu_min = GPU.getAdjustedFreqs(this).toTypedArray()
        builder.setItems(gpu_min) { dialog, which ->
            GPU.setMinFreq(GPU.getAvailableFreqs()[which] / 1000000, this@gpuActivity)
            Prefs.saveInt("gpu_min_freq", GPU.getAvailableFreqs()[which] / 1000000, this@gpuActivity)
            refreshUI()
        }
        val dialog = builder.create()
        dialog.show()
    }

    private fun maxDialog() {
        val builder = AlertDialog.Builder(ContextThemeWrapper(this@gpuActivity, R.style.CustomDialogTheme))
        builder.setTitle("GPU maximum freq")
        val gpu_max = GPU.getAdjustedFreqs(this).toTypedArray()
        builder.setItems(gpu_max) { dialog, which ->
            GPU.setMaxFreq(GPU.getAvailableFreqs()[which], this@gpuActivity)
            Prefs.saveInt("gpu_max_freq", GPU.getAvailableFreqs()[which], this@gpuActivity)
            refreshUI()
        }
        val dialog = builder.create()
        dialog.show()
    }

    fun pwrDialog() {
        val builder = AlertDialog.Builder(ContextThemeWrapper(this@gpuActivity, R.style.CustomDialogTheme))
        builder.setTitle("GPU power level")
        val mpLayout = layoutInflater.inflate(R.layout.gpu_pwr_dialog, null)
        val editText = mpLayout.findViewById<EditText>(R.id.pwr_lvl_et)
        editText.setText(GPU.getPwrLevel())
        builder.setPositiveButton("Apply") { dialog, which ->
            GPU.setPwrLevel(editText.text.toString(), this@gpuActivity)
            Prefs.saveString("gpu_pwr_lvl", editText.text.toString(), this@gpuActivity)
            refreshUI()
        }
        builder.setView(mpLayout)
        val dialog = builder.create()
        dialog.show()
    }
}
